internal/controller/management/projects/projects.go

```diff
@@ -392,6 +400,74 @@ func (r *reconciler) ensureAPIAdminPermissions(
 	return nil
 }
 
+func (r *reconciler) ensureControllerPermissions(
+	ctx context.Context,
+	project *kargoapi.Project,
+) error {
+	logger := logging.LoggerFromContext(ctx).WithValues(
+		"project", project.Name,
+		"namespace", project.Name,
+	)
+
+	// Get all ServiceAccounts labeled as controller ServiceAccounts
+	controllerSAs := &corev1.ServiceAccountList{}
+	if err := r.client.List(ctx, controllerSAs, client.InNamespace(r.cfg.KargoNamespace), client.MatchingLabels{"app.kubernetes.io/component": "controller"}); err != nil {
+		return fmt.Errorf("error listing controller ServiceAccounts: %w", err)
+	}
+
+	// Create/update a RoleBinding for each ServiceAccount
+	for _, sa := range controllerSAs.Items {
+		roleBindingName := fmt.Sprintf("kargo-controller-secrets-readonly-%s", sa.Name)
+		logger := logger.WithValues("serviceAccount", sa.Name, "roleBinding", roleBindingName)
+
+		roleBinding := &rbacv1.RoleBinding{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      roleBindingName,
+				Namespace: project.Name,
+			},
+			RoleRef: rbacv1.RoleRef{
+				APIGroup: rbacv1.GroupName,
+				Kind:     "ClusterRole",
+				Name:     "kargo-controller-secrets-readonly",
+			},
+			Subjects: []rbacv1.Subject{
+				{
+					Kind:      "ServiceAccount",
+					Name:      sa.Name,
+					Namespace: r.cfg.KargoNamespace,
+				},
+			},
+		}
+
+		if err := r.createRoleBindingFn(ctx, roleBinding); err != nil {
+			if kubeerr.IsAlreadyExists(err) {
+				existingRoleBinding := &rbacv1.RoleBinding{}
+				if err := r.client.Get(ctx, client.ObjectKey{Name: roleBindingName, Namespace: project.Name}, existingRoleBinding); err != nil {
+					return fmt.Errorf("error getting existing RoleBinding: %w", err)
+				}
+				existingRoleBinding.Subjects = roleBinding.Subjects
```

Comment:
I think I would not bother to retrieve the existing RB. I would probably jump straight to updating using the RB struct you've already built. There isn't any justifiable reason that anyone should be tampering with this RB as it belongs to _Kargo_ and not to a user. It should unconditionally look like what we say, so imho, we don't need to be careful about not squashing any modifications a user may have made to the existing RB.

Note that _if_ Projects were more frequently reconciled, I would probably encourage a different approach here, BUT as things currently stand, Projects are only reconciled until they are fully initialized or fail to initialize (without possibility of recovering). This means, on average, each Project is reconciled only once. Creating the RB will be the thing that usually needs to happen here, so its sensible to continue leading off with that, then dealing with the very unlikely case that the RB already existed by updating it, as I said, unconditionally.